policy-controller/k8s/index/src/pod.rs

```diff
@@ -68,6 +68,52 @@ pub(crate) fn tcp_port_names(spec: Option<k8s::PodSpec>) -> HashMap<String, Port
     port_names
 }
 
+/// Gets the container probe ports for a Pod.
+///
+/// The result is a mapping for each probe port exposed by a container in the
+/// Pod and the paths for which probes are expected.
+pub(crate) fn get_container_probes(
+    spec: &Option<k8s::PodSpec>,
+    port_names: &HashMap<String, PortSet>,
+) -> PortMap<HashSet<String>> {
+    let mut probes = PortMap::<HashSet<String>>::default();
+    for container in spec.iter().flat_map(|s| s.containers.iter()) {
+        set_probe_ports(&container.liveness_probe, &mut probes, port_names);
+        set_probe_ports(&container.readiness_probe, &mut probes, port_names);
+        set_probe_ports(&container.startup_probe, &mut probes, port_names);
+    }
+    probes
+}
+
+fn set_probe_ports(
+    probe: &Option<k8s::Probe>,
+    probes: &mut PortMap<HashSet<String>>,
+    port_names: &HashMap<String, PortSet>,
+) {
+    let probe = match probe {
+        Some(probe) => probe,
+        None => return,
+    };
+    if let Some(http) = &probe.http_get {
+        if let Some(path) = &http.path {
+            match &http.port {
+                k8s::IntOrString::Int(port) => {
+                    if let Ok(port) = u16::try_from(*port).and_then(NonZeroU16::try_from) {
+                        let paths = probes.entry(port).or_default();
+                        paths.insert(path.clone());
+                    }
+                }
+                k8s::IntOrString::String(name) => {
+                    for port in port_names.get(name).into_iter().flatten() {
+                        let paths = probes.entry(*port).or_default();
+                        paths.insert(path.clone());
+                    }
+                }
+            }
+        }
+    }
+}
```

Comment:
I don't love passing in `&mut probes` when we're just trying to construct a list of probes. This, to me, indicates that we're managing probes and may be removing items or otherwise mutating it.

We can make a single pipeline for this with a little modification:

```suggestion
    let mut probes = PortMap::<HashSet<String>>::default();
    for (port, path) in spec
        .iter()
        .flat_map(|s| s.containers.iter())
        .flat_map(|c| {
            (c.liveness_probe.iter())
                .chain(c.readiness_probe.iter())
                .chain(c.startup_probe.iter())
        })
        .flat_map(|p| get_http_probes(p, port_names).into_iter().flatten())
    {
        probes.entry(port).or_default().insert(path);
    }
    probes
}

fn get_http_probes(
    probe: &k8s::Probe,
    port_names: &HashMap<String, PortSet>,
) -> Option<Vec<(NonZeroU16, String)>> {
    let http = probe.http_get.as_ref()?;
    let path = http.path.clone().unwrap_or_else(|| "/".to_string());
    let ports = match &http.port {
        k8s::IntOrString::Int(port) => {
            let port = u16::try_from(*port).and_then(NonZeroU16::try_from).ok()?;
            vec![(port, path)]
        }
        k8s::IntOrString::String(name) => port_names
            .get(name)?
            .iter()
            .copied()
            .map(|p| (p, path.clone()))
            .collect::<Vec<_>>(),
    };
    debug_assert!(!ports.is_empty(), "probe ports list must not be empty");
    Some(ports)
}
```

This also avoids passing in an `Option<&Probe>` to the function which is reading a probe, etc.